File: src/clu/input.py
"""Doc Incomplete."""

import logging
from pathlib import Path
import shlex
import shutil
import subprocess
from typing import Optional, Union

log = logging.getLogger(__name__)

FILE_SIZE_LIMIT = 1 * 1024 * 1024  # 1MB - arbitrary limit - seems reasonable


def text_file(fname: Union[str, Path], optional: bool = False) -> str:
    """Read a file and return its contents."""
    log.debug(f"text_file: {fname=}")

    fname = Path(fname)
    if not fname.is_file():
        if optional:
            log.info(f"Optional file not found: {fname}")
        else:
            log.error(f"File not found: {fname}")
        return ""

    file_size = fname.stat().st_size
    log.debug(f"File size is for {fname}: {file_size}")
    if file_size > FILE_SIZE_LIMIT:
        log.error(f"File size exceeds limit: {file_size} > {FILE_SIZE_LIMIT}")
        return ""

    # TODO: check permissions

    with open(fname, "r") as f:
        return f.read()


def text_program(cmdline) -> tuple[str, int]:
    log.debug(f"text_program: {cmdline}")

    # TODO: check for existence FIRST
    # if not check_program_exists(cmdline):
    #     log.error(f"Program not found: {cmdline}")
    #     return "", 1

    # TODO: check for permissions

    # TODO: check for output size

    try:
        result = subprocess.run(shlex.split(cmdline), capture_output=True, text=True)
        return result.stdout, result.returncode
    except Exception as e:
        log.error(f"Error running program {cmdline}: {e}")
        return "", 1


def transform_cmdline_to_filename(cmdline: str) -> tuple[str, str]:
    # cmdline is space separated, so we need to convert spaces to underscores
    cmdline = cmdline.replace(" ", "_")

    # udevadm info uses path like things that are not really paths - get rid of slashes
    cmdline = cmdline.replace("/", "%")

    log.debug(f"transformed {cmdline=}")
    return cmdline, cmdline + "_rc"


def check_program_exists(program: str) -> Optional[str]:
    log.debug(f"check_program_exists: {program}")

    return shutil.which(program.split()[0])  # Only check the actual command, not its arguments


def check_file_exists(fname: Union[str, Path]) -> Optional[str]:
    log.debug(f"check_file_exists: {fname}")

    fname = Path(fname)
    exists = fname.is_file()
    if exists:
        log.debug(f"File {fname} found")
        return str(fname)
    else:
        log.debug(f"File {fname} not found")
        return None

File: pkg/input/input.go
package input

// Port of src/clu/input.py providing utilities for reading text files and running
// simple text-based programs, plus helper transformations.

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

const OutputSizeLimit = 1 * 1024 * 1024 // 1MB arbitrary limit

// TextProgram runs a command line (split on spaces) and returns combined
// stdout/stderr output, the exit code, and an error. If the process runs and
// exits with a non-zero status the returned error will be of type
// *exec.ExitError (so callers can inspect it), and the exit code will be
// returned as the second value.
func TextProgram(cmdline string) (string, int, error) {
	parts := strings.Fields(cmdline)
	if len(parts) == 0 {
		return "", 1, fmt.Errorf("empty command line")
	}
	cmd := exec.Command(parts[0], parts[1:]...)
	out, err := cmd.CombinedOutput()
	output := string(out)

	// Derive an exit code we can return even if the process failed.
	exitCode := 0
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			exitCode = ee.ExitCode()
		} else {
			exitCode = 1
		}
	}

	// Enforce a maximum output size similar to TextFile's FileSizeLimit.
	// BUG: we need to stop at the limit while reading output, not after.
	if len(out) > OutputSizeLimit {
		return output, exitCode, fmt.Errorf("output size exceeds limit: %d > %d", len(out), OutputSizeLimit)
	}

	return output, 0, nil
}

// TransformCmdlineToFilename converts a command line into a safe-ish filename.
// Spaces -> underscores, slashes -> percent signs. Returns the filename and
// the filename with _rc suffix (akin to Python returning tuple of name and rc name).
func TransformCmdlineToFilename(cmdline string) (string, string) {
	name := strings.ReplaceAll(cmdline, " ", "_")
	name = strings.ReplaceAll(name, "/", "%")
	return name, name + "_rc"
}

// CheckProgramExists returns the absolute path to the program if found in PATH,
// otherwise empty string.
func CheckProgramExists(program string) string {
	first := strings.Fields(program)
	if len(first) == 0 {
		return ""
	}
	path, err := exec.LookPath(first[0])
	if err != nil {
		return ""
	}
	return path
}

// CheckFileExists returns the absolute path if the file exists, else empty string.
func CheckFileExists(fname string) string {
	abs, err := filepath.Abs(fname)
	if err != nil {
		return ""
	}
	info, err := os.Stat(abs)
	if err != nil || info.IsDir() {
		return ""
	}
	return abs
}

// TextFile reads a text file and returns its contents as a string. If the file
// does not exist or cannot be read, an error is returned. If the file is empty
// an error is also returned. Files larger than OutputSizeLimit are rejected.
func TextFile(fname string) (string, error) {
	info, err := os.Stat(fname)
	if err != nil {
		if os.IsNotExist(err) {
			return "", os.ErrNotExist
		}
		return "", fmt.Errorf("stat %s: %w", fname, err)
	}

	if info.IsDir() {
		return "", fmt.Errorf("path is a directory: %s", fname)
	}

	if info.Size() > OutputSizeLimit {
		return "", fmt.Errorf("file size exceeds limit: %d > %d", info.Size(), OutputSizeLimit)
	}

	data, err := os.ReadFile(fname)
	if err != nil {
		return "", fmt.Errorf("read %s: %w", fname, err)
	}

	// FIXME: consider whether empty files are truly an error in all contexts
	// FIXME: base this off the stat call instead of reading the file?
	if len(data) == 0 {
		return "", fmt.Errorf("empty file content")
	}
	return string(data), nil
}
